Several updates on progress bar and early warning

truebelief opened the pull request:
Hi @dgirardeau (and @rjanvier),

I’ve just pushed a new commit ([ed5969a](https://github.com/truebelief/cc-treeiso-plugin/commit/ed5969a7ac9f9513daa703d33eee0e4b89d6375d)) with trackable changes:

1. Simplified code for progress updates:
A callback function has been used to update the progress bar.

2. Finer progress increments:
The progress bar now increments in smaller steps.

3. Rapid ground detection:
A quick check based on height quantile is now added before the initial segmentation step to spot potential ground points.

4. Warning before initial segmentation:
Users now get a prompt before proceeding. They can opt to quit without running the initial segmentation if the data contains ground points.

Could I ask you to test these updates again? Your previous feedback was really helpful in screening the weaknesses

truebelief committed:
Major upgrade of dependencies and running time

1. Removed the Boost dependency by adopting a newer cut-pursuit version from the authors (resulting in a 5–15× speed improvement).
2. Fixed an issue with the nearest neighbor search caused by an incorrect 2D array order input to the knn_cpp query function.
3. Enabled customization of the number of threads for knn-cpp, yielding a slight speed improvement.
4. Optimized array calculations through enhanced use of STL algorithms.

truebelief committed:
Change log added

truebelief committed:
Smoothened the progress bar and added ground detectioned

- Improved progress bar increment
- Added rapid detection of ground points based on quantile distribution
- Added warning of ground points before continuing initial segmentation

dgirardeau commented:
I've just tested it.

The progress bar is indeed more reactive (even if it still stops way before 100%).

And the ground detection seems to work. But the plugin still crashes if you proceed 😉

truebelief commented:
> I've just tested it.
> 
> The progress bar is indeed more reactive (even if it still stops way before 100%).
> 
> And the ground detection seems to work. But the plugin still crashes if you proceed 😉


Thanks for making it clearer! Now I know I should test a larger point cloud with heavy ground coverage. I'll do some debugging on that and see if I can track down the crash. Will get back to you

dgirardeau commented:
Do you need the file I used?

truebelief commented:
> Do you need the file I used?

That would be great! I just added you as the editor of shared google drive folder with your gmail, if that makes it easier to upload

dgirardeau commented:
I've uploaded a file named 5054801_c.las (somewhere in the Spanish Basque country).

![image](https://github.com/user-attachments/assets/5127fdb0-5843-41e9-abd3-b4122af8a081)

I like this file because it has some stray points floating in the air 😆 . They are a worst case for a lot of point vicinity-based algorithms.

truebelief commented:
> I've uploaded a file named 5054801_c.las (somewhere in the Spanish Basque country).
> 
> ![image](https://private-user-images.githubusercontent.com/2735509/413064662-5127fdb0-5843-41e9-abd3-b4122af8a081.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0ODE0ODEsIm5iZiI6MTczOTQ4MTE4MSwicGF0aCI6Ii8yNzM1NTA5LzQxMzA2NDY2Mi01MTI3ZmRiMC01ODQzLTQxZTktYWJkMy1iNDEyMmFmOGEwODEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTNUMjExMzAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmJjODg0N2VmYzNmNjQwOGZjZDgwMTAxOTc3YWNmYThmYjQ2ODA4MmUzOWZlNWQ4OWQzNWEyNWEwNWI1ZmNiMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.3BvAHLB9LRemU10rlbT79C6wVtdRnjv-VFMOR46fIxg)
> 
> I like this file because it has some stray points floating in the air 😆 . They are a worst case for a lot of point vicinity-based algorithms.

Yes, those points did trip up my tree algorithm. There are also power line points in the mix. Thanks for sharing! I'll take a closer look when I’m back home.

truebelief committed:
Crash fix

- Avoided crashes by adding error handling once the maximum component limit is reached
- Improved the speed of unique and get_subset functions

truebelief commented:
I fixed the crash issue, which happened because the program was exiting improperly once new segments exceeded the uint16_t (65535) limit. Now it exits with an error popup instead of silently crashing.

I also updated the unique function to use a hash table for better performance. You can check out the new commit here: [7d4e8cd1](https://github.com/truebelief/cc-treeiso-plugin/pull/10/commits/7d4e8cd1e8e6a49100292d87e805c25a9246818c). Please let me know how it works for you!

It might be tedious to review every single edit. If everything tests out fine, I can go ahead and close this PR and merge the branch into the main. Thanks!

dgirardeau commented:
I confirm that it doesn't crash anymore.

dgirardeau commented:
Not sure what the conflicts are though. You may want to fix that before merging?

truebelief committed:
Merge branch 'main' into optimize_treeiso_speed

truebelief commented:
> Not sure what the conflicts are though. You may want to fix that before merging?

Thanks for taking time. I just corrected the conflicts and merged it. Finger crossed!

truebelief commented:
Thank you again for updating the plugin to the alpha version!
